De-flake chat-room rate-limit test: 100 msg/sec window races ack-drain wall-clock

### Task
The test `rate-limits beyond 100 msg/sec/room with RATE_LIMITED` in `apps/worker/src/__tests__/chat-room.test.ts` is wall-clock-sensitive and flakes on loaded CI runners. It fires 100 `send` frames, then drains 100 acks in an `await ws.next(2000)` loop, then sends a 101st and asserts it returns `error`/`RATE_LIMITED`. The `RollingWindowLimiter` uses real `Date.now()` with a 1000ms window, so when the ack-drain loop takes longer than the window (slow runner), the earliest entries have aged out by the time the 101st send is evaluated — it is allowed, the assertion gets `'ack'` instead of `'error'`, and the job fails. After this issue, the test must deterministically exercise the limit regardless of runner speed (no wall-clock race).

### Context
This flake has caused false `lint · typecheck · test` failures on at least PR #37 (forced a rebase + re-run during 2026-05-16 backlog triage) and PR #53 (passed only on re-run). It taxes every PR that touches the worker until fixed: a green re-run is currently the only mitigation, which masks real failures and burns CI. Flagged during PR-triage on 2026-05-16.

### Reproduction
1. Run `apps/worker` vitest on a loaded machine (or CI under contention).
2. The 100-ack drain loop in the test exceeds ~1000ms wall-clock.
3. **Expected:** 101st send → `{ kind: "error", code: "RATE_LIMITED" }`.
4. **Actual:** rolling window has expired entries; 101st send → `{ kind: "ack" }`; `AssertionError: expected 'ack' to be 'error'` at `chat-room.test.ts:184`.
- Environment: GitHub Actions `lint · typecheck · test`; intermittent. Passes locally and on unloaded runs.

### Pointers
- `apps/worker/src/__tests__/chat-room.test.ts:163` — the flaky `it("rate-limits beyond 100 msg/sec/room ...")` block; ack-drain loop ~line 176, failing assertion ~line 184
- `apps/worker/src/do/ChatRoom.ts:57` — `RATE_LIMIT_WINDOW_MS = 1000`, `RATE_LIMIT_PER_WINDOW = 100`
- `apps/worker/src/do/ChatRoom.ts:122` — `new RollingWindowLimiter(RATE_LIMIT_PER_WINDOW, RATE_LIMIT_WINDOW_MS)`
- `apps/worker/src/do/ChatRoom.ts:280` — `this.limiter.tryAcquire(Date.now())` (real clock; the racing call)
- `apps/worker/src/lib/rate-limit.ts` — `RollingWindowLimiter` implementation (`tryAcquire(now)` already takes an explicit timestamp)
- Prior occurrences: [PR #37](https://github.com/schmug/loomwiki/pull/37), [PR #53](https://github.com/schmug/loomwiki/pull/53)

### Constraints
- Keep production behavior unchanged — this is a test/limiter-seam fix only; do not weaken the real 100 msg/sec/room cap.
- No arbitrary `sleep`/timeout inflation as the "fix" — the test must be deterministic, not just slower.
- Workers runtime only via `@cloudflare/vitest-pool-workers`; no Node-specific timer APIs.
- `tryAcquire(now)` already accepts an injected timestamp — prefer leveraging that seam over rewriting the limiter.

### Acceptance criteria
- The test no longer depends on the 100-ack drain completing within the 1000ms window (e.g. submit all 101 sends before draining, OR inject a controllable clock so the window is held constant across the assertion).
- Running the worker suite under artificial load / repeated runs (≥20 iterations) shows 0 failures of this test.
- The test still genuinely asserts the 101st in-window send is rejected with `RATE_LIMITED` (not trivially weakened to always pass).
- `pnpm test` in `apps/worker` passes; no new flakiness elsewhere.

### Out of scope
- Changing the rate-limit policy (window size / cap).
- The `blockConcurrencyWhile` rollback skip in `chat-room-atomicity.test.ts`.
- Broader test-suite flakiness audit beyond this one test.

--- a/apps/worker/src/__tests__/chat-room.test.ts
+++ b/apps/worker/src/__tests__/chat-room.test.ts
@@ -4,9 +4,10 @@
 // route + DO upgrade all run as one unit. Reserves runInDurableObject for
 // cases the route layer cannot reach (alarm fire, mirror failure paths).
 
-import { SELF, env } from "cloudflare:test";
+import { SELF, env, runInDurableObject } from "cloudflare:test";
 import { PROTOCOL_VERSION } from "@loomwiki/shared";
 import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import type { ChatRoom } from "../do/ChatRoom.js";
 import { applyMigrations, resetDb } from "./__fixtures__/db.js";
 import { type JwtFixture, makeJwtFixture } from "./__fixtures__/jwt.js";
 import { type WsSession, openWs } from "./__fixtures__/ws.js";
@@ -162,6 +163,20 @@ describe("ChatRoom DO — WebSocket flow", () => {
 
   it("rate-limits beyond 100 msg/sec/room with RATE_LIMITED", async () => {
     const alice = await bootstrapMember("alice@example.com", "general");
+
+    // Freeze the rate-limiter clock via the _nowFn seam so the 1-second
+    // rolling window never advances during the ack-drain loop. Without this,
+    // the ack-drain (100 × `await ws.next()`) can exceed 1000ms wall-clock on
+    // a loaded CI runner, causing the window to expire and the 101st send to
+    // be allowed instead of rejected — the original flake (#55).
+    const frozenNow = Date.now();
+    const stub = (env.CHAT_ROOM as DurableObjectNamespace<ChatRoom>).get(
+      (env.CHAT_ROOM as DurableObjectNamespace<ChatRoom>).idFromName(alice.roomId),
+    );
+    await runInDurableObject(stub, (instance: ChatRoom) => {
+      instance._nowFn = () => frozenNow;
+    });
+
     const ws = await track(await openWs(alice.roomId, alice.jwt));
     ws.send({ kind: "hello", protocolVersion: PROTOCOL_VERSION });
     await ws.next();
@@ -170,15 +185,16 @@ describe("ChatRoom DO — WebSocket flow", () => {
     for (let i = 0; i < 100; i++) {
       ws.send({ kind: "send", tempId: `t-${i}`, body: `m${i}` });
     }
-    // Drain acks (and any echoed messages — there's only one socket so we
-    // expect 100 acks total).
+    // Drain acks — the frozen clock guarantees all 100 hits remain in window
+    // no matter how long this loop takes in real time.
     let acks = 0;
     while (acks < 100) {
       const m = await ws.next(2000);
       if (m.kind === "ack") acks++;
     }
 
-    // The 101st must be rejected.
+    // The 101st must be rejected because the frozen clock keeps all 100 prior
+    // hits inside the window.
     ws.send({ kind: "send", tempId: "rate", body: "rate-me" });
     const next = await ws.next();
     expect(next.kind).toBe("error");

--- a/apps/worker/src/do/ChatRoom.ts
+++ b/apps/worker/src/do/ChatRoom.ts
@@ -115,6 +115,14 @@ export class ChatRoom extends DurableObject<Env> {
    */
   _mirrorFn: typeof mirrorMessageToD1 | undefined = undefined;
 
+  /**
+   * Optional clock override for the rate limiter. Defaults to `undefined`
+   * (production uses `Date.now()`). Tests can inject a frozen timestamp via
+   * `runInDurableObject` so the rolling window never advances during the
+   * ack-drain loop, making the rate-limit test deterministic (#55).
+   */
+  _nowFn: (() => number) | undefined = undefined;
+
   constructor(ctx: DurableObjectState, env: Env) {
     super(ctx, env);
     this.sql = ctx.storage.sql;
@@ -277,7 +285,7 @@ export class ChatRoom extends DurableObject<Env> {
       sendError(ws, ErrorCodes.VALIDATION_FAILED, "body exceeds 4096 chars", env.tempId);
       return;
     }
-    if (!this.limiter.tryAcquire(Date.now())) {
+    if (!this.limiter.tryAcquire((this._nowFn ?? Date.now.bind(Date))())) {
       sendError(ws, ErrorCodes.RATE_LIMITED, "100 msg/sec/room cap reached", env.tempId);
       return;
     }

